galette/lib/Galette/Entity/Contribution.php

```diff
@@ -263,41 +262,46 @@ private function retrieveEndDate(): void
         $begin_date = new \DateTime($this->begin_date);
         if ($preferences->pref_beg_membership != '') {
             //case beginning of membership
-            list($j, $m) = explode('/', $preferences->pref_beg_membership);
-            $next_begin_date = new DateTime($begin_date->format('Y') . '-' . $m . '-' . $j);
-            while ($next_begin_date <= $begin_date) {
-                $next_begin_date->add(new \DateInterval('P1Y'));
-            }
+            if ($this->type->extension < 1) { // 0 = donate; -1= else duration (months)
+                list($j, $m) = explode('/', $preferences->pref_beg_membership);
+                $next_begin_date = new DateTime($begin_date->format('Y') . '-' . $m . '-' . $j);
+                while ($next_begin_date <= $begin_date) {
+                    $next_begin_date->add(new DateInterval('P1Y'));
+                }
 
-            if ($preferences->pref_membership_offermonths > 0) {
-                //count days until next membership begin date
-                $diff1 = (int)$now->diff($next_begin_date)->format('%a');
+                if ($preferences->pref_membership_offermonths > 0) {
+                    //count days until next membership begin date
+                    $diff1 = (int)$now->diff($next_begin_date)->format('%a');
 
-                //count days between next membership begin date and offered months
-                $tdate = clone $next_begin_date;
-                $tdate->sub(new \DateInterval('P' . $preferences->pref_membership_offermonths . 'M'));
-                $diff2 = (int)$next_begin_date->diff($tdate)->format('%a');
+                    //count days between next membership begin date and offered months
+                    $tdate = clone $next_begin_date;
+                    $tdate->sub(new DateInterval('P' . $preferences->pref_membership_offermonths . 'M'));
+                    $diff2 = (int)$next_begin_date->diff($tdate)->format('%a');
 
-                //when number of days until next membership begin date is less than or equal to the offered months, it's free :)
-                if ($diff1 <= $diff2) {
-                    $next_begin_date->add(new \DateInterval('P1Y'));
+                    //when number of days until next membership begin date is less than or equal to the offered months, it's free :)
+                    if ($diff1 <= $diff2) {
+                        $next_begin_date->add(new DateInterval('P1Y'));
+                    }
                 }
-            }
 
-            // Caution : the end_date to retrieve is the day before the next_begin_date.
-            $end_date = clone $next_begin_date;
-            $end_date->sub(new \DateInterval('P1D'));
-            $this->end_date = $end_date->format('Y-m-d');
```

Comment:
`$this->end_date` seems no longer set

Reply:
Fixed. That part is not covered in current tests :cry:

Reply:
OK, I found the issue. This part should has been tested, but since switching to phpunit, expcetions are no longer handled the same way, and should be the last instruction.